src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1716,10 +1716,11 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                             .cannot_borrow_path_as_mutable(span, &item_msg, Origin::Mir);
                         err.span_label(span, "cannot borrow as mutable");
 
-                        if place != place_err {
-                            if let Some(name) = self.describe_place(place_err) {
-                                err.note(&format!("the value which is causing this path not to be \
-                                    mutable is...: `{}`", name));
+                        if let &Place::Local(local) = place_err {
+                            let local_decl = &self.mir.local_decls[local];
+                            if local_decl.is_user_variable && local_decl.mutability == Mutability::Not {
+                                err.span_label(local_decl.source_info.span,
+                                               format!("consider changing this to `mut {}`", local_decl.name.unwrap()));
                             }
                         }
```

Comment:
Also, I was [noting on Zulip at 4:40pm (ET) yesterday](https://rust-lang.zulipchat.com/#narrow/stream/122657-wg-nll/subject/diagnostics/near/127382282), I think we want to move this suggestion code so that it comes *after* the match. That will require restructuring this method somewhat. The problem with the current setup is that your suggestion will only fire when you are trying to  *borrow* something, but not — I think — in tests like `let x = 3; x = 4;` (try it!). I would combine it with some of the [other suggestion-like code in the method too][s], yielding something like this:

```rust
let mut err = match kind {
  Reservation(...) | Write(...) => {
    if no error { return } else { construct_err_object() }
  }
  Reservation(...) | Write(...) => /* some variant of the preivous arm */,
  Reservation(...) | Write(...) => ...,
};

// If we get here, then there *is* an error
match place_err {
    Place::Local(l) => {
        // suggest changing to `mut` variable
    }
    Place::Projection(...) => {
        // suggest changing to `&mut` borrow (see [link 1])
    }
}

err.emit()
```

[link 1][s]

[s]: (https://github.com/rust-lang/rust/blob/61f35e507a56dcdce88bfce99bb2d1eeacb0e9d1/src/librustc_mir/borrow_check/mod.rs#L1723-L1728)

Reply:
We are also going to want to watch out for "upvars" — those are cases where the `place_err` will not be a `Place::Local` but we want to make a similar suggestion. We can discuss on Zulip though, I'm just going to edit your PR description to have a check-list so things don't get forgotten.